Go fit chi2

gtaillepied opened the pull request:
Chi2 without covariance matrix.

Commit:
hubble-fit: diagrams, fit (broken)

Copy of the Python code for everything but the covariance matrix (and
associated calculations). Hubble diagram's fit gives an incorrect omgM.
Left to do : fix fit, cosmetics on the diagrams, cov matrix.

Commit:
hubble-fit: no covar matrix

working version, chi2 minimization done without covariance matrix.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
-// Copyright 2017 The hubble-fit Authors.  All rights reserved.
-// Use of this source code is governed by a BSD-style
+// Copyright 2017 The hubble-fit authors. All rights reserved.
```
please keep the spelling as originally written.
(ie: `Authors`)

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
-// Copyright 2017 The hubble-fit Authors.  All rights reserved.
-// Use of this source code is governed by a BSD-style
+// Copyright 2017 The hubble-fit authors. All rights reserved.
+// Use of this source code is governed by a BSD-style 
```
please remove the lingering space at the end.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
 
-// hubble-fit is a program to fit the Hubble diagram with a Cosmological model.
+// hubble-fit is a program to fit the Hubble diagram with a Cosmological model
+
```
please remove that empty line (otherwise, `godoc` won't associate that comment with the documentation for that package/command.)

also, please keep the dot `.` at the end of the (rather succint) doc line.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
 package main
 
+import (
```
it's customary to group the imports by: "standard library imports" and then "third party imports".

ie:

```go
import (
 "fmt"
 "image/color"
 ...
 "strings"

 "github.com/gonum/integrate"
 ...
 "go-hep.org/x/hep/fit"
)
```

this is automatically performed by `goimports`.
e.g.:

```sh
$> go get -u -v golang.org/x/tools/cmd/goimports
$> cd hubble-fit
$> goimports -w .
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+	// compute the difference between two slices, value by value
+
+	var diff_slice []float64
```
as you know the final length of these 2 slices, it's better to pre-allocate them with that size.

ie:
```go
diff := make([]float64, len(s1))
abs := make([]float64, len(s1))

for i := range s1 {
    val := s2[i] - s1[i]
    diff[i] = val
    abs[i] = math.Abs(val)
}
return diff, abs
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	// inputs : z = redshift, n = number of steps for the trapeze method
+
+	var inte float64
+	abs := make([]float64, 0)
```
same comment here: you know the number of iterations (`n+1`) and thus the final size of the `abs` and `ord` slices.
please create them with the appropriate size (and adapt the for-loops below accordingly)

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	// inputs : z = redshift, n = number of steps for the integral, to be passed to integral function
+
+	var mu_th float64
+	mu_th = 5 * math.Log10(((1+z)*c/(10*H))*integral(z, n))
```
these 2 lines could easily be replaced with just:

```go
mu_th := 5 * math.Log10(((1+z)*c/(10*H))*integral(z, n))
```
or even:

```go
// mu_th returns the theoretical mu for each supernova, given 
// the redshift z and the number of steps n for the integral calculation.
func mu_th(z, n float64) float64 {
    return 5 * math.Log10(((1+z)*c/(10*H))*integral(z, n))
}
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	return mu_th
+}
+
+func mu_th_slice(s []float64, n float64) []float64 {
```
same comments than for `func mu_th`

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+######
+*/
+
+func mu_exp_slice(m_stell, mb, stretch, colour []float64) []float64 {
```
same comments than for `func mu_th`

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	return slice
+}
+
+func mu_exp_err(dmb, dstretch, dcolour []float64) []float64 {
```
same comments than for `func mu_th`

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+	slice := make([]float64, 0)
+	for i := 0; i < len(dmb); i++ {
+		val := math.Sqrt(math.Pow(dmb[i], 2) + math.Pow(alpha*dstretch[i], 2) + math.Pow(beta*dcolour[i], 2))
```
`math.Pow` isn't supposed to be used for such trivial powers (ie: it's not very fast because it needs to be able to handle floating point powers...)

better create 3 temporary values and multiply:

```go
slice := make([]float64, len(dmb))
for i := range dmb {
  ai := dmb[i]
  bi := alpha*dstretch[i]
  ci := beta*dcoulour[i]
  slice[i] = math.Sqrt(ai*ai + bi*bi + ci*ci)
}
return slice
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+######
+*/
+
+func points(x_values, y_values []float64) plotter.XYs {
```
idiomatic Go would name `x_values` into `xs` and `y_values` into `ys`.

also, moving the commented lines below just before the `func points` declaration would make the function properly documented.

I would also probably rename `func points` as `func newPoints` to make the intend explicit.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	return pts
+}
+
+func diag(points plotter.XYs) *plot.Plot {
```
`diag` is a bit ambiguous as a function name: is it short for `diagonal`?

maybe just call it `hubblePlot` ? or `newHubblePlot`?

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+		ord = append(ord, val)
+	}
+
+	inte = integrate.Trapezoidal(abs, ord)
```
just return it directly without using this temporary variable that you had trouble naming :)

```go
return integrate.Trapezoidal(abs, ord)
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	// inputs : z = redshift, omega = omgM as a free parameter
+
+	var inte float64
+	abs := make([]float64, 0)
```
please create the `abs` and `ord` slices with their final size.

also, instead of `abs` and `ord` that are shortcuts for abscissa and ordinate, just name those `xs` and `ys`.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	ord := make([]float64, 0)
+
+	var i float64
+	for i = 0; i <= 1000; i++ {
```
it's probably possible (and faster) to loop only once:

```go
const N=1000
xs := make([]float64, N+1)
ys := make([]float64, N+1)
for i := range xs {
   x := z * float64(i)/float64(N)
   xs[i] = x
   ys[i] = 1 / math.Sqrt((1+x*omega)*(1+x)*(1+x)-(1-omega)*(2+x)*x)
}
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+}
+
+func histogram (s []float64) *plot.Plot {
```
rename `histogram` into `newHistogram` ?

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	}
+
+	val := make(plotter.Values, len(s))
+	for i,_ := range s {
```
that can be simplified into:

```go
for i := range s {
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+	// creates slices with the needed values from JLA datas
+
+	sn_names := make([]string, 0)	// names
```
you know _a priori_ the final lengths of these slices. just create them with it:

```go
var (
    N = len(supernovae)
    sn_names = make([]string, N) // names of the supernovae
    zcmb = make([]float64, N)    // redshift in the CMB frame
    ...
)
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+			case 19:
+				z, _ := strconv.ParseFloat(y, 64)
+				de_jla = append(de_jla, z)
+			}
```
you may want to add a `default` case to detect for misformated or corrupted data files.
alternatively, check for the size of `split_str` (instead of adding that `default` case.)
then you can drop the `switch` and directly index into `split_str`

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	// compute and store slices needed for the plots and analysis
+
+	var muexp_data []float64		// experimental mus from SALT2 model
+	muexp_data = mu_exp_slice(m_stell, mb, stretch, colour)
```
you can merge these 2 lines into just one (but keep the comment):

```go
muexp_data := mu_exp_slice(m_stell, mb, stretch, colour) // experimental mus from SALT2 model
```

ditto for the slices below.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+*/
+	// compute, draw and store the various diagrams
+
+	scatter_data := points(zcmb, muexp_data)
```
loose the temporary variable `scatter_data` and just call `points` inside `diag`:
```go
hubblePlot := diag(points(zcmb, muexp_data))
err = hubblePlot.Save(1080, 720, "hubble_diagram.png")
if err != nil {
    log.Fatalf("error saving 'hubble_diagram.png': %v", err)
}
```

ditto for `scatter2_data`.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+	scatter2_data := points(zcmb, diff_data)
+	diagram2 := diag(scatter2_data)
+	diagram2.Save(720, 200, "hubble_diagram2.png")
```
check for a non `nil` error returned by `Save`.
you do want to be sure that what you get on your disk isn't corrupted or is instead the plot from last execution.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+
+	empty := make([]float64, len(zcmb))
+	params := make([]float64, 0)
+	params = append(params, 0.295, 0.141, 3.101, -19.05, -0.070)
```
this can be re-written as:

```go
params := []float64{0.295, 0.141, 3.101, -19.05, -0.070}
```

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	empty := make([]float64, len(zcmb))
+	params := make([]float64, 0)
+	params = append(params, 0.295, 0.141, 3.101, -19.05, -0.070)
+	res,_ := fit.Curve1D(
```
please check for a non=`nil` error returned by `Curve1D`.
you do want to be sure that what you print out as a fit result isn't some garbage (it may very well be if an error occured).

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+						index = p
+					}
+				}
+				if m_stell[index] < 10 {
```
these 4 lines could be simplified as:
```go
out := mb[index]-ps[3]+ps[1]*stretch[index]-ps[2]*colour[index] - 5*math.Log10(((1+z)*c/(10*H))*modified_integral(z, ps[0]))
if m_stell[index] >= 10 {
    out -= ps[4]
}
return out
```

also, `index` is usually spelled out `i`.

sbinet commented on main.go:
```diff
@@ -1,9 +1,482 @@
+	res,_ := fit.Curve1D(
+		fit.Func1D{
+			F: func(z float64, ps []float64) float64 {
+				var index int
```
I would initialize `index` to `-1` to detect if something bad happened (e.g. the floating point comparison goes astray, or if we didn't manage to find a correct index)
and check that `index >= 0` after the for-loop and panic with some meaningful error message.

sbinet reviewed (commented):
looks good (only looked at the code, not the physics results, though), modulo the few comments I have reported.

Commit:
hubble-fit: whithout covar matrix

corrected version, with simplifications and cosmetics

sbinet commented on main.go:
```diff
@@ -1,9 +1,513 @@
 // license that can be found in the LICENSE file.
-
 // hubble-fit is a program to fit the Hubble diagram with a Cosmological model.
+
```
no, actually, it's the line above that needed to be kept and this one to be dropped.
ie:

```go
// Copyright 2017 The hubble-fit Authors.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// hubble-fit is a program to fit the Hubble diagram with a Cosmological model.
package main

```

Commit:
hubble-fit: no covariance matrix

useless commented sections deleted

sbinet commented on main.go:
```diff
@@ -1,9 +1,441 @@
+	abs := make([]float64, int(n+1))
+	ord := make([]float64, int(n+1))
+
+	for i := 0; i <= int(n); i++ {
```
these 2 loops could be combined into 1:

```go
zn := z / n
for i := range abs {
    x := zn * float64(i)
    y := 1 / math.Sqrt((1+x*omgM)*(1+x)(1+x)-(1-omgM)*(2+x)*x)
    abs[i] = x
    ord[i] = y
}
```

also, better rename `abs` and `ord` as `xs` and `ys`.

sbinet commented on main.go:
```diff
@@ -1,9 +1,441 @@
+
+	diag_ord := make([]float64, len(s))
+
+	for i := 0; i < len(s); i++ {
```
or just:
```go
for i := range s {
...
```

Commit:
hubble-fit: with covar matrix (broken)

broken implementation of a chi2 test with covariance matrix
pushed for reviewing

Commit:
hubble-fit: chi2 with covariance matrix, no fit

rewritten chi2 function to compute it only from the list of varying
parameters

Commit:
hubble-fit: chi2 with covar matrix, broken

version including a test function to compute the five parameters through
a chi2 minimization that accounts for the covariance matrix. outputs
wrong values.

sbinet reviewed (approved):
LGTM.

could you please send another pull request (separate from this one) adding yourself to the `AUTHORS` and `CONTRIBUTORS` files ?

Once that's done, I'll merge this in.

thanks.

sbinet commented:
ping?

sbinet commented:
A+C handled via 707fc5e

sbinet commented:
fixed via lsst-lpc/hubble-fit#4.
